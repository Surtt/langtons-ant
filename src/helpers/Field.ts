type Cell = false | true;
type Line = Cell[];
export type Field = Line[];

export const generateField = (size: number, state: Cell): Field =>
  new Array<boolean[]>(size)
    .fill(new Array<boolean>(size))
    .map(() => new Array<boolean>(size).fill(state));

// export const toggleCell = (cell: Cell, coords: Coords): Cell => { };
