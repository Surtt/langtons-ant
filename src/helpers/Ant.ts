import { Cell } from '@/helpers/Field';
export type Coords = [number, number];

type RandomGenerator<R = number> = (
  min: number,
  max: number,
  rotation?: number
) => R;

export type Ant = {
  xAnt: number;
  yAnt: number;
  rotation: number;
};

export const generateNumber: RandomGenerator = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const generateAntPosition: RandomGenerator<Ant> = (
  min: number,
  max: number,
  rotation = 0
) => {
  const xAnt = generateNumber(min, max);
  const yAnt = generateNumber(min, max);
  return {
    xAnt,
    yAnt,
    rotation,
  };
};

export const changeAntState = (
  rotation: number,
  cell: Cell
): {
  xAnt: number;
  yAnt: number;
} => {
  const value = { xAnt: 0, yAnt: 0 };
  switch (rotation) {
    case 90:
      value.xAnt++;
      break;
    case 180:
      value.yAnt--;
      break;
    case 270:
      value.xAnt--;
      break;
    default:
      value.yAnt++;
      break;
  }
  if (cell) {
    value.xAnt = -value.xAnt;
    value.yAnt = -value.yAnt;
  }
  return value;
};

export const newRotation = (rotation: number, cell: Cell): number => {
  let newRotation = rotation + 90;
  if (cell) {
    newRotation += 180;
  }
  if (newRotation >= 360) {
    newRotation -= 360;
  }
  return newRotation;
};
