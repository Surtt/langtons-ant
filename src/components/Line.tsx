import React, { FC } from 'react';
import { Cell } from './Cell';
import { Coords } from '../helpers/Ant';

interface LineProps {
  line: boolean[];
  index: number;
  ant: Coords;
}

export const Line: FC<LineProps> = ({ line, index, ant }) => {
  const [x, y] = ant;
  return (
    <tr className="line">
      {line.map((cell, idx) => (
        <Cell
          key={`line_${index}-cell_${idx}`}
          ant={index === y && idx === x ? ant : undefined}
        />
      ))}
    </tr>
  );
};
