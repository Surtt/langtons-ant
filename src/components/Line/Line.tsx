import React, { FC } from 'react';
import { Cell } from '../Cell/Cell';
import { Ant } from '@/helpers/Ant';
import { Line as LineType } from '../../helpers/Field';

export interface LineProps {
  /**
   * Line data
   */
  line: LineType;
  /**
   * Line index
   */
  index: number;
  /**
   * Ant position
   */
  ant: Ant;
}

export const Line: FC<LineProps> = ({ line, index, ant }) => {
  const { xAnt, yAnt } = ant;
  return (
    <tr>
      {line.map((cell, idx) => (
        <Cell
          key={`line_${index}-cell_${idx}`}
          ant={index === xAnt && idx === yAnt ? ant : undefined}
        />
      ))}
    </tr>
  );
};
