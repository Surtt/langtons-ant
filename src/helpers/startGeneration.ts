import { changeAntState, newRotation, Ant } from './Ant';
import { Field, toggleCellOnTheField } from './Field';
export interface GeneratingValues {
  field: Field;
  ant: Ant;
  count: number;
}

const isFieldBorder = (field: Field, ant: Ant) => {
  const borderY = field.length - 1;
  const borderX = field[0].length - 1;
  return (
    ant.yAnt === 0 ||
    ant.yAnt === borderY ||
    ant.xAnt === 0 ||
    ant.xAnt === borderX
  );
};

export const startGeneration = (
  field: Field,
  ant: Ant,
  count: number
): GeneratingValues => {
  const cnt = count + 1;
  const { xAnt, yAnt, rotation } = ant;

  const movement = changeAntState(rotation, field[xAnt][yAnt]);
  const updateRotation = newRotation(rotation, field[xAnt][yAnt]);
  field = toggleCellOnTheField(field, [xAnt, yAnt]);
  movement.xAnt += xAnt;
  movement.yAnt += yAnt;
  const newAnt = {
    xAnt: movement.xAnt,
    yAnt: movement.yAnt,
    rotation: updateRotation,
  };

  if (isFieldBorder(field, ant)) {
    const length = field.length;
    field.unshift(new Array<boolean>(length).fill(false));
    field.push(new Array<boolean>(length).fill(false));
    field.forEach((item) => {
      item.unshift(false);
      item.push(false);
    });
    if (newAnt) {
      newAnt.xAnt++;
      newAnt.yAnt++;
    }
  }

  return {
    field: [...field],
    ant: newAnt,
    count: cnt,
  };
};
