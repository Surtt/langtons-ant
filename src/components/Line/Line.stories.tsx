import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Line, LineProps } from './Line';
import { BoardTable } from '../Board/Board';

export default {
  title: 'Board/Line',
  component: Line,
  decorators: [
    (Story) => (
      <BoardTable style={{ width: '15rem' }}>
        <Story />
      </BoardTable>
    ),
  ],
} as Meta;

const Template: Story<LineProps> = (args) => <Line {...args} />;

export const LineExample = Template.bind({});
LineExample.args = {
  line: [false, false, false, false, false, false, false, false, false, false],
  index: 0,
  ant: { xAnt: 0, yAnt: 4, rotation: 0 },
};
