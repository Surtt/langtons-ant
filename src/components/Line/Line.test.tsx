import React from 'react';
import { render, screen } from '@testing-library/react';
import { Line } from './Line';
import { Line as LineType } from '@/helpers/Field';

const line: LineType = [false, false, false, false, false];
const index = 0;
const ant = {
  xAnt: 0,
  yAnt: 0,
  rotation: 0,
};

it('Renders Line', () => {
  const { asFragment } = render(<Line line={line} index={index} ant={ant} />);
  expect(asFragment()).toMatchSnapshot();
});

it('renders Line role', () => {
  render(<Line line={line} index={index} ant={ant} />);
  expect(screen.getByRole('line')).toBeInTheDocument();
});
