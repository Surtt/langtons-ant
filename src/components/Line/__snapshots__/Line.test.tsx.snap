// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Line 1`] = `
<DocumentFragment>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-10u676d"
    />
  </div>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-bdbrua"
    />
  </div>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-bdbrua"
    />
  </div>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-bdbrua"
    />
  </div>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-bdbrua"
    />
  </div>
</DocumentFragment>
`;
